File: 3. Longest Substring Without Repeating Characters/solution.cs
public class Solution
{
    public int LengthOfLongestSubstring(string s)
    {
        var set = new HashSet<char>();
        var longest = 0;
        var current = 0;
        var subIdx = 0;

        foreach (var r in s)
        {
            if (set.Contains(r))
            {
                if (current > longest)
                {
                    longest = current;
                }

                for (; subIdx < s.Length; subIdx++)
                {
                    set.Remove(s[subIdx]);
                    current--;
                    if (s[subIdx] == r)
                    {
                        break;
                    }
                }

                subIdx++;
            }

            current++;
            set.Add(r);
        }

        if (current > longest)
        {
            longest = current;
        }

        return longest;
    }
}

File: 3. Longest Substring Without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        have: set[str] = set()
        longest = 0
        current = 0
        sub_idx = -1

        for c in s:
            if c in have:
                if current > longest:
                    longest = current

                while True:
                    sub_idx += 1
                    have.remove(s[sub_idx])
                    current -= 1
                    if s[sub_idx] == c:
                        break

            current += 1
            have.add(c)

        if current > longest:
            longest = current

        return longest
